### ISSUE TITLE
No type mismatch for records with enum keys inside object literal with dynamic key

### ISSUE THREAD

**Issue by droooney** [microsoft/TypeScript#63655]
### 🔎 Search Terms

enum dynamic key record

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about records

### ⏯ Playground Link

https://www.typescriptlang.org/play/?ts=6.0.3#code/C4TwDgpgBAglC8UAKBDATsAligNgHgCUIBjAezQBM8BpCEAGigDsBXAWwCMI0A+HgbgBQg0JCi0QCKAHIU0qAB8ZHaUMFkmAZ2BQA5hGASpACgCUALnF0EPGXLUbtUFJbiIA3oKhQA2vsN0ZgC6ltLaaJhMutL0ggC+-EA

### 💻 Code

```ts
type A = Partial<Record<Key, number>>;

type Key = 'a' | 'b';

const getKey = (): Key => 'a';

const a: A = {
  [getKey()]: 'string',
};
```

### 🙁 Actual behavior

TS didn't report an error even though it should

### 🙂 Expected behavior

Error report about type mismatch inside the record field

### Additional information about the issue

As far as I've investigated it doesn't happen when the key is not an enum (i.e. `string`). Also when the dynamic key is narrowed down to a single option the error does get reported. However, when there are at least 2 options no error is reported.

**Comment by MartinJohns** [comment]
Duplicate of #38663.

Note: there is no [enum](https://www.typescriptlang.org/docs/handbook/enums.html) in your example. `'a' | 'b'` is called a [union type](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types).

**Comment by droooney** [comment]
> Duplicate of [#38663](https://github.com/microsoft/TypeScript/issues/38663).
> 
> Note: there is no [enum](https://www.typescriptlang.org/docs/handbook/enums.html) in your example. `'a' | 'b'` is called a [union type](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types).

My bad, indeed it is a union, I mixed up the names. However, [the same thing happens with enums](https://www.typescriptlang.org/play/?ts=6.0.3#code/C4TwDgpgBAglC8UAKBDATsAligNgHgCUIBjAezQBM8BpCEAGigDsBXAWwCMI0A+HgbgBQgiKzZRaIKAG9BUKCgRQA5CmX05UDkuUd1ggL7CyTAM7AoAcwjBJSgBQBKAFwS6CHm5AA6FEMEm5gqucIiy8gDa1rZ0TgC6rsrmaJhMlvoG-EA):

```ts
type A = Partial<Record<Key, number>>;

enum Key {
  a = 'a',
  b = 'b',
}

const getKey = (): Key => Key.a;

const a: A = {
  [getKey()]: 'string',
};
```
